[FLINK-28227][connectors] Migrate o.a.f.streaming.examples to the new Source API

afedulov opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
Migrates the existing examples to use the new `DataGeneratorSource` instead of `SourceFunction` and reimplements the `SocketSource(Function)` as the new Source . This change is required to unblock https://github.com/apache/flink/pull/20049/files (CI checks do not allow usage of deprecated APIs in examples).


## Verifying this change

This change is covered by the existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

afedulov commented:
@flinkbot run azure

reswqa commented:
Hi @afedulov, would you minding telling me how it is progressing now, thanks.

afedulov commented:
Hi @reswqa I actually just finished reimplementing the `SocketSource(Function)`  as the new Source and - this unblocks https://github.com/apache/flink/pull/20049 . I just need to go though some internal OSS contributions formality and will most likely complement this PR with last missing bit early next week.

afedulov commented:
@reswqa I just pushed the latest changes, please take a look.

afedulov commented:
@flinkbot run azure

afedulov commented:
@flinkbot run azure

afedulov commented:
@flinkbot run azure

reswqa commented on flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/connectors/SocketSource.java:
```diff
@@ -0,0 +1,235 @@
+            String hostname,
+            int port,
+            byte byteDelimiter,
+            DeserializationSchema<RowData> deserializer) {
```
I'm not aware where we opened it but this should be necessary for some `DeserializationSchema`.

afedulov replied:
I will add an open call analogous to how `KafkaSource` handles it:
https://github.com/apache/flink/blob/9e8a99c12ce939418086fa9555c1c4f74bcf6b59/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSource.java#L144

reswqa commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/iteration/IterateExample.java:
```diff
@@ -96,7 +100,21 @@ public static void main(String[] args) throws Exception {
+                            dataGenerator,
+                            BOUND,
+                            RateLimiterStrategy.perSecond(20),
+                            TypeInformation.of(new TypeHint<Tuple2<Integer, Integer>>() {}));
```
Could `Types.TUPLE(Types.INT, Types.INT)` work here? If possible, it would be more concise.

reswqa reviewed (commented):
Thanks @afedulov for the migration! I left some comments, please take a look in your free time.

afedulov commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/async/AsyncIOExample.java:
```diff
@@ -76,8 +79,19 @@ public static void main(String[] args) throws Exception {
+                env.fromSource(
+                        generatorSource,
+                        WatermarkStrategy.noWatermarks(),
+                        "Generated tuples Source");
```
Thanks, a copy-paste error.

afedulov commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/TopSpeedWindowing.java:
```diff
@@ -99,7 +104,20 @@ public static void main(String[] args) throws Exception {
+            carData =
+                    env.fromSource(
+                            carGeneratorSource,
+                            WatermarkStrategy.noWatermarks(),
```
Yes, and the watermarks are assigned here https://github.com/apache/flink/pull/21774/files#diff-227d83380c24406389c12d4d3116ced03c15493d638b029595431051695c2549L109 because of this if-else block
https://github.com/apache/flink/pull/21774/files#diff-227d83380c24406389c12d4d3116ced03c15493d638b029595431051695c2549L86 and the file source which parses data in a separate map function. There is some consensus that this example does not really serve its purpose, so I don't think it makes sense to put too much effort into "fixing" it: https://github.com/apache/flink/pull/20049#issuecomment-1407553971

reswqa replied:
> There is consensus that this example does not really serve its purpose, so I don't think it makes sense to put much effort into "fixing" it, it is enough that it produces data and compiles.

As for this `TopSpeedWindowing` example, it is fair enough as I'm also not a fan of this. Perhaps we should show users a more intuitive example that cover the core logic of this in the future.

afedulov replied:
To be honest I am not quite sure which feature this example originally intended to cover. I know that there is also some consensus that usage of advanced windows with triggers and evictors is somewhat discouraged and should not be promoted. 
cc @alpinegizmo would appreciate your opinion

reswqa replied:
> There is also some consensus that usage of advanced windows with triggers and evictors is somewhat discouraged and should not be promoted.
Agree on this point.  Whether removing or improving this example, it should not block this migrate ATM.

afedulov commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/SessionWindowing.java:
```diff
@@ -64,24 +67,19 @@ public static void main(String[] args) throws Exception {
-                            public void run(SourceContext<Tuple3<String, Long, Integer>> ctx)
-                                    throws Exception {
-                                for (Tuple3<String, Long, Integer> value : input) {
-                                    ctx.collectWithTimestamp(value, value.f1);
```
Good point, I think I accidentally missed that detail when I worked on this part of the PR a while back. I'll add a timestamp extractor. Doing this on the level of GeneratorFunction is not possible - that would require some redesign around a low-level implementation with user-supplied SerDe, similar to how KafkaSource does it: https://github.com/apache/flink/blob/8db43d7ffeabc7a0f5c67c9cdd964c64f400789b/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaRecordEmitter.java#L60

afedulov commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/TopSpeedWindowing.java:
```diff
@@ -99,7 +104,20 @@ public static void main(String[] args) throws Exception {
+                    new DataGeneratorSource<>(
+                            carGenerator,
+                            Long.MAX_VALUE,
+                            parallelismIgnored -> new GuavaRateLimiter(1),
```
To be honest it is hard to reason how the data is supposed to be generated by the original `CarSource` to begin with. It updates the distance as if one second had passed between the events, but it actually loops with 100ms delay, which is overestimating the distance by 10. Then there are these magic numbers of triggerMeters and eviction:
https://github.com/apache/flink/blob/af4e0506ce39e014d56a28d36cfc497bf0aa00f0/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/TopSpeedWindowing.java#L123-L124
Since the new `CarSource` does not overestimate the distance I will reduce the delta window threshold and increase the events rate - this seems to produce the results at the same rate.

afedulov commented:
@reswqa Thanks for the thorough review! 
I addressed your comments in a separate commit [`96d0a93` (#21774)](https://github.com/apache/flink/pull/21774/commits/96d0a936547da8818584838ac35661663ff2af7a)

reswqa reviewed (commented):
Thanks for the update, it seems that CI is failed related to this change.

afedulov commented:
Arghhh, this example is broken 🙄 
@reswqa The issue is that the original function miscalculates the distance by making it 10 times larger than it is supposed to be based on the time elapsed between sampling (described here https://github.com/apache/flink/pull/21774#discussion_r1242618518, basically it thinks the speed is 10 times larger, not 0-100km/h, but 0-1000km/h - some next gen cars). 
This threshold https://github.com/apache/flink/blob/af4e0506ce39e014d56a28d36cfc497bf0aa00f0/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/TopSpeedWindowing.java#L124 controls how often the window trigger will fire, so in the old example it feels that the data is generated "faster" simply because the cars all move super fast, as you originally correctly pointed out in https://github.com/apache/flink/pull/21774#discussion_r1239993143.

So there are three potential solutions I see to fix the test:
1) Keep the `triggerMeters=50`, but overestimate the speed by a factor of 10, same way as the original function did (I checked - when the distance is simply multiplied by, the 10 results come out much faster and also the test passes)
2) Keep the correct distance calculation and modify the threshold to `triggerMeters=5`. This will lead to the same rate of produced results, but it is the most cumbersome solution, because the test data needs to be fixed by adding about 100 samples "manually": https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/util/TopSpeedWindowingExampleData.java#L23
3) Keep the correct distance calculation, but also keep the `triggerMeters=50` - this will manifest itself as the "slower" data generation, there are simply not too many samples that exceed the delta when the calculation is done right.

Given how broken this example is, I would personally prefer to go for option 3) since it is the most "correct" one making too many changes. That said, I am open to consider the alternatives if you think otherwise.

reswqa commented:
Well, from my point of view, how to generate data here is not what users really want to care about. We just need to maintain a self consistent logic. So It is ok for me to choose both option (1) and (3).

reswqa reviewed (approved):
@afedulov Thanks for the update, look good to me. Please don't forget to remove the tmp commit and squash all other fix-up commits. I will merge this after that assuming AZP passes.

afedulov commented:
@flinkbot run azure
